crates/iceberg-catalog/src/catalog/io.rs

```diff
@@ -1,14 +1,16 @@
 use crate::api::{ErrorModel, Result};
 use crate::service::storage::path_utils;
 use flate2::{write::GzEncoder, Compression};
-use iceberg::io::FileIO;
+use iceberg::{io::FileIO, spec::view_properties::METADATA_COMPRESSION};
 use iceberg_ext::catalog::rest::IcebergErrorResponse;
 use serde::Serialize;
 use std::io::Write;
 
+use super::CommonMetadata;
+
 pub(crate) async fn write_metadata_file(
     metadata_location: &str,
-    table_metadata: impl Serialize,
+    metadata: impl CommonMetadata + Serialize,
```

Comment:
Let's pass the compression codec into here so we don't have to verify properties in this function. I'm thinking something along the lines of:


```rust

#[derive(thiserror::Error, Debug)]
#[error("Unsupported compression codec: {0}")]
pub struct UnsupportedCompression(pub String);

#[derive(Debug, Clone, Copy)]
pub enum CompressionCodecs {
    Gzip,
    None,
}

impl CompressionCodecs {
    pub fn compress(&self, payload: &[u8]) -> Result<Vec<u8>, IoError> {
        match self {
            CompressionCodecs::Gzip => {
                let mut compressed_metadata = GzEncoder::new(Vec::new(), Compression::default());
                compressed_metadata.write_all(payload).map_err(IoError::FileCompression)?;
                compressed_metadata.finish().map_err(IoError::FileCompression)
            }
            CompressionCodecs::None => {
                Ok(payload.to_vec())
            }
        }
    }

    pub fn as_file_extension(&self) -> &'static str {
        match self {
            CompressionCodecs::Gzip => "gz",
            CompressionCodecs::None => "",
        }
    }
}

impl CompressionCodecs {
    pub fn try_from_metadata(val: impl CommonMetadata) -> Result<Self, UnsupportedCompression> {
        val.properties()
            .get(METADATA_COMPRESSION)
            .map(String::as_str)
            .map_or(Ok(CompressionCodecs::Gzip), |codec| {
                match codec {
                    "gzip"  => Ok(CompressionCodecs::Gzip),
                    "none" => Ok(CompressionCodecs::None),
                    _ => Err(UnsupportedCompression(codec.to_string())),
                }
            })
    }
}

pub(crate) async fn write_metadata_file(
    metadata_location: &str,
    metadata: impl Serialize,
    codec: CompressionCodecs,
    file_io: &FileIO,
) -> Result<(), IoError> {
...
}
```

That way, we can avoid passing the untyped (`HashMap<String, String>`) properties into the location fns in `crates/iceberg-catalog/src/service/storage/` and remove the duplication of functionality, thanks to the 
`as_file_extension` fn we can easily figure out what ending to use without having to deny unsupported codecs again over there.

By introducing another function on the CommonMetadata trait:

```rust
pub(crate) trait CommonMetadata {
    fn properties(&self) -> &HashMap<String, String>;
    fn compression(&self) -> Result<io::CompressionCodecs, UnsupportedCompression> {
        Ok(io::CompressionCodecs::try_from_metadata(self)?)
    }
}
```

we could then just extract the compression codec in the respective handler fns and short-circuit there. E.g. in

Reply:
Thanks for that suggesion. I was hesitant to indroduce "even more" new structs or enums, but I like your suggestion much more than passing in a HashMap.